### DISCUSSION TITLE
Finding two isomorphic Subgraphs

### DISCUSSION THREAD

**Discussion by thomasahle** [networkx/networkx#7517] [Q&A]
What is the best way to identify isomorphic subgraphs in a networkx graph?

If I already knew one of the subgraphs, it would be easy to do with the graphmatcher/vf2 algorthm.
But I don't.

If course, in simple graphs there might be a lot of small isomorphic subgraphs (like every individual node). I'm only interested in isomorphic subgraphs above a certain size.

**Comment by dschult** [comment]
Hmmm...  If I understand correctly, a brute force solution would be to iterate over all subgraphs above a certain size and look for isomorphic copies of each subgraph in the network. Is this correct?

But the wording "subgraph" an "isomorphic subgraph" are subtle and small changes in definition can make a big difference. As I've stated the problem above, this sounds like a slow process. But there may be other approaches. What is the actual problem you are trying to solve by using this method? There may be a way to view that problem that avoids much of the computation.

**Comment by Peiffap** [comment]
You might be able to use the modular product by reducing the task of finding subgraph isomorphisms to the (equally computationally complex, but perhaps easier to reason about) task of finding cliques in the modular product graph.

**Comment by thomasahle** [comment]
@dschult You are right - the actual problem I'm working on has somewhat more structure.
I'm using this for [tensorgrad](https://github.com/thomasahle/tensorgrad) and I need to find isomorphic subgraphs that share all their edges:
<img width="742" alt="Screenshot 2024-06-28 at 12 32 08 PM" src="https://github.com/networkx/networkx/assets/946355/125ca57d-5e3b-4204-982b-d5e5c237e65c">
The tricky part is that I also support hyper-edges (which I turn into small white nodes in the picture), so in practice they two subgraphs  may only be connected through a separator of such white nodes.

The graphs I'm interested in don't typically have a lot of isomorphism. Most nodes have different labels, which of course speeds vf2 up a lot. It's not like I'm working on huge regular graphs of blank nodes.
So I thought maybe just iterating all isomorphic subgraphs would be fine - assuming there was some pruned vf2 way to do it - and then check the connection criterion afterwards.

@Peiffap The modular product seems cool!
Does `nx. find_cliques` have a way to specify "semantic equivalence" the same way GraphMatcher does? (https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.isomorphism.GraphMatcher.semantic_feasibility.html#networkx.algorithms.isomorphism.GraphMatcher.semantic_feasibility)

**Reply by Peiffap** [reply]
I'm not aware of a built-in way of specifying semantic equivalence, but I could be wrong about that. I imagine you could hack something together to post-process the cliques after you find them, though. Note that `nx.find_cliques()` only returns _maximal_ cliques; you might want to look at [`nx.enumerate_all_cliques()`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.clique.enumerate_all_cliques.html#networkx.algorithms.clique.enumerate_all_cliques) instead.

**Comment by thomasahle** [comment]
There's also the [McSplit algorithm](https://www.ijcai.org/proceedings/2017/0099.pdf) (see also Wikipedia: [Maximum common induced subgraph](https://en.wikipedia.org/wiki/Maximum_common_induced_subgraph). Could this be interesting to integrate in Networkx?
There's a [python implementation here](https://gitlab.insa-rouen.fr/lulmer/asi5_louisulmer/-/tree/e4b5730f638ff4f43c509c1b4fce4566378dc7d0/bibleS9/GraphAnalysis/Python_Maximum_Common_Induced_Subgraph-master).

**Reply by dschult** [reply]
Be careful about using other people's code when writing code for NetworkX. We need the original author to submit the code (or authorize the submission of code). And if you read that code in detail it will be very hard for you to implement a version that isn't a variant of what they did. We need to preserve and recognize code authorship and license.

Before we pursue that, did you try the McSplit algorithm or that code for your problem? Does it satisfy what you need?

**Reply by thomasahle** [reply]
I wasn't planning on adding it to NetworkX. I just wanted to suggest it could be relevant. For my project I think I'll end up having to write my own branch and bound algorthm.

**Comment by HansBrende** [comment]
@thomasahle wouldn't `networkx.algorithms.isomorphism.ISMAGS.largest_common_subgraph` be what you're looking for or am I missing something?
